pkg/workloadmeta/store.go

```diff
@@ -0,0 +1,375 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package workloadmeta
+
+import (
+	"context"
+	"sync"
+	"time"
+
+	"github.com/DataDog/datadog-agent/pkg/errors"
+	"github.com/DataDog/datadog-agent/pkg/status/health"
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/retry"
+)
+
+var globalStore *Store
+
+const (
+	retryCollectorInterval = 30 * time.Second
+	pullCollectorInterval  = 5 * time.Second
+	eventBundleChTimeout   = 1 * time.Second
+	eventChBufferSize      = 50
+)
+
+type subscriber struct {
+	name   string
+	ch     chan EventBundle
+	filter *Filter
+}
+
+// Store contains metadata for workloads.
+type Store struct {
+	storeMut sync.RWMutex
+	store    map[Kind]map[string]Entity
+
+	subscribersMut sync.RWMutex
+	subscribers    []subscriber
+
+	candidates map[string]Collector
+	collectors map[string]Collector
+
+	eventCh chan []Event
+}
+
+// NewStore creates a new workload metadata store. Call Run to start it.
+func NewStore() *Store {
+	candidates := make(map[string]Collector)
+	for id, c := range collectorCatalog {
+		candidates[id] = c()
+	}
+
+	return &Store{
+		store:       make(map[Kind]map[string]Entity),
+		subscribers: []subscriber{},
+
+		candidates: candidates,
+		collectors: make(map[string]Collector),
+		eventCh:    make(chan []Event, eventChBufferSize),
+	}
+}
+
+// Run starts the workload metadata store.
+func (s *Store) Run(ctx context.Context) {
+	retryTicker := time.NewTicker(retryCollectorInterval)
+	pullTicker := time.NewTicker(pullCollectorInterval)
+	health := health.RegisterLiveness("workloadmeta-store")
+
+	// Start collectors immediately
+	s.startCandidates(ctx)
+
+	// Start a pull immediately to fill the store without waiting for the
+	// next tick.
+	pullCtx, pullCancel := context.WithTimeout(ctx, pullCollectorInterval)
+	s.pull(pullCtx)
+
+	log.Info("workloadmeta store initialized successfully")
+
+	go func() {
+		for {
+			select {
+			case <-health.C:
+
+			case <-pullTicker.C:
+				// pullCtx will always be expired at this point
+				// if pullTicker has the same duration as
+				// pullCtx, so we cancel just as good practice
+				pullCancel()
+
+				pullCtx, pullCancel = context.WithTimeout(ctx, pullCollectorInterval)
+				s.pull(pullCtx)
+
+			case evs := <-s.eventCh:
+				s.handleEvents(evs)
+
+			case <-retryTicker.C:
+				s.startCandidates(ctx)
+
+				if len(s.candidates) == 0 {
+					retryTicker.Stop()
+				}
+
+			case <-ctx.Done():
+				retryTicker.Stop()
+				pullTicker.Stop()
+
+				err := health.Deregister()
+				if err != nil {
+					log.Warnf("error de-registering health check: %s", err)
+				}
+
+				return
+			}
+		}
+	}()
+}
+
+// Subscribe returns a channel where workload metadata events will be streamed
+// as they happen.
```

Comment:
This function appears to also generate some "synthetic" eventa for existing entities.  That should probably be in the docstring as well.